addon/components/abstract-modal.js

```diff
@@ -0,0 +1,146 @@
+
+    // pass in the target DOM node, as well as the observer options
+    mutationObserver.observe(this.get('$containerEl').get(0), config)
+
+    return mutationObserver
+  },
+
+  /**
+  * update perfect-scrollbar before checking scoll position
+  * need to disconnect MutationObserver then re-establish to prevent infinite loop
+  * of mutation event triggers, callback updates, which triggers new mutation event...
+  */
+  updatePerfectScrollbar () {
+    let mutationObserver
+    const $containerEl = this.get('$containerEl')
+
+    Ps.update($containerEl.get(0))
+
+    if (this.get('containerObserver')) {
+      this.get('containerObserver').disconnect()
+      mutationObserver = this.createMutationObserver()
+    }
+
+    if (this.isDestroyed || this.isDestroying) {
+      return
+    } else {
+      this.set('containerObserver', mutationObserver)
+    }
+  },
+
+  /**
+  * Checks scroll position within container element to add/remove scroll styling to header/footer elements
+  */
+  updateScrollStyles () {
+    const $containerEl = this.get('$containerEl')
+    const $headerEl = this.get('$headerEl')
+    const $footerEl = this.get('$footerEl')
+    const scrollTop = $containerEl.scrollTop()
+    const innerHeight = $containerEl.innerHeight()
+    const scrollHeight = $containerEl.get(0) && $containerEl.get(0).scrollHeight
+
+    this.updatePerfectScrollbar()
+
+    // style top of form with box shadow if content overflow hidden underneath header
+    if (scrollTop > 0) {
+      $headerEl.addClass('header-scrolled')
+    } else {
+      $headerEl.removeClass('header-scrolled')
+    }
+
+    // style bottom of form with box shadow if content overflow hidden underneath footer
+    if (scrollTop + innerHeight >= scrollHeight) {
+      $footerEl.removeClass('footer-scrolled')
+    } else {
+      $footerEl.addClass('footer-scrolled')
+    }
+  },
+
+  /**
+   * sets bindings/callbacks for perfect-scrollbar events on document element
+   * creates a window.MutationObserver for mutating events on container element
+   */
+  setScrollBindings () {
+    const $headerEl = this.get('$headerEl')
+    const $footerEl = this.get('$footerEl')
+
+    this.set('containerObserver', this.createMutationObserver())
+
+    // bind document to perfect-scrollbar events
+    $(document).on('ps-scroll-up', () => $footerEl.addClass('footer-scrolled'))
+    $(document).on('ps-y-reach-end', () => $footerEl.removeClass('footer-scrolled'))
+    $(document).on('ps-scroll-down', () => $headerEl.addClass('header-scrolled'))
+    $(document).on('ps-y-reach-start', () => $headerEl.removeClass('header-scrolled'))
+
+    $(window).resize(this.updateScrollStyles.bind(this))
+
+    // update scroll styles in case where content already hidden on initial DOM render
+    this.updateScrollStyles()
+
+    this.set('scrollBindingsSet', true)
+  },
+  /* Ember.Component method */
+  willDestroyElement () {
```

Comment:
But I don't see where it is disconnection the mutation observer created in `createMutationObserver()`.